File: android/src/main/java/org/ligi/gobandroid_hd/logic/cell_gatherer/MustBeConnectedCellGatherer.java
package org.ligi.gobandroid_hd.logic.cell_gatherer;

import org.ligi.gobandroid_hd.logic.StatefulGoBoard;
import org.ligi.gobandroid_hd.logic.StatelessBoardCell;

public class MustBeConnectedCellGatherer extends CellGatherer {


    public MustBeConnectedCellGatherer(StatefulGoBoard board, StatelessBoardCell root) {
        super(board, root);
    }

    @Override
    protected void pushWithCheck(StatelessBoardCell cell) {
        final boolean unProcessed = processed.add(cell);

        if (board.areCellsEqual(cell, root)) {
            add(cell);
            if (unProcessed) {
                pushSurroundingWithCheck(cell);
            }
        }
    }
}

File: android/src/main/java/org/ligi/gobandroid_hd/logic/cell_gatherer/MustBeConnectedCellGatherer.kt
package org.ligi.gobandroid_hd.logic.cell_gatherer

import org.ligi.gobandroid_hd.logic.StatefulGoBoard
import org.ligi.gobandroid_hd.logic.StatelessBoardCell

open class MustBeConnectedCellGatherer(board: StatefulGoBoard, root: StatelessBoardCell) : CellGatherer(board, root) {


    override fun process(cell: StatelessBoardCell) {
        if (processed.add(cell) && board.areCellsEqual(cell, root)) {
            gatheredCells.add(cell)

            cell.neighbors.forEach {
                process(it)
            }

        }
    }

}
